File: 0064-minimum-path-sum/0064-minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        mat=[[0]*len(grid[0]) for i in range(len(grid))]
        mat[0][0]=grid[0][0]
        for i in range(1,len(mat)):
            mat[i][0]=mat[i-1][0]+grid[i][0]
        for j in range(1,len(mat[0])):
            mat[0][j]=mat[0][j-1]+grid[0][j]
        for i in range(1,len(grid)):
            for j in range(1,len(grid[0])):
                mat[i][j]=min(mat[i-1][j],mat[i][j-1])+grid[i][j]
        return mat[-1][-1]

File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int [][] dp = new int [grid.length][grid[0].length];
        int n = grid.length;
        int m = grid[0].length;
        dp[n-1][m-1] = grid[n-1][m-1];
        
        for(int i=n-2;i>=0;i--)
            dp[i][m-1] = grid[i][m-1]+dp[i+1][m-1];
        
        for(int i=m-2;i>=0;i--)
            dp[n-1][i] = grid[n-1][i]+dp[n-1][i+1];

        for(int i=n-2;i>=0;i--) {
            for(int j = m-2;j>=0;j--) {
                dp[i][j] = grid[i][j] + Math.min(dp[i][j+1],dp[i+1][j]);
            }
        }
        return dp[0][0];
    }
}